src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets

```
@@ -2079,8 +2081,61 @@ because xbuild doesn't support framework reference assemblies.
 	</_BuildApkEmbedInputs>
 </PropertyGroup>
 
-<Target Name="_BuildApkEmbed"
+<!--
+  Find all assemblies that need to be compressed and set up the @(_AssembliesToCompress) item
+  group so that we can run an MSBuild incremental target to compress them.
+-->
+<Target Name="_CollectAssembliesToCompress"
   DependsOnTargets="_PrepareBuildApk"
+  Condition="'$(EmbedAssembliesIntoApk)' == 'True'">
+
+  <CollectAssemblyFilesToCompress
+      AssemblyCompressionDirectory="$(IntermediateOutputPath)android\lz4\"
+      EmbedAssemblies="$(EmbedAssembliesIntoApk)"
+      EnableCompression="$(AndroidEnableAssemblyCompression)"
+      IncludeDebugSymbols="$(AndroidIncludeDebugSymbols)"
+      ProjectFullPath="$(MSBuildProjectFullPath)"
+      ResolvedFrameworkAssemblies="@(_ShrunkFrameworkAssemblies)"
+      ResolvedUserAssemblies="@(_ShrunkUserAssemblies);@(_AndroidResolvedSatellitePaths)"
+      SupportedAbis="@(_BuildTargetAbis)">
+    <Output TaskParameter="AssembliesToCompressOutput" ItemName="_AssembliesToCompress" />
+    <Output TaskParameter="ResolvedFrameworkAssembliesOutput" ItemName="_CompressedResolvedFrameworkAssemblies" />
+    <Output TaskParameter="ResolvedUserAssembliesOutput" ItemName="_CompressedBuildApkResolvedUserAssemblies" />
+  </CollectAssemblyFilesToCompress>
+</Target>
+
+<!--
+  LZ4 compress all compressable assemblies. Note this is an incremental MSBuild target.
+  @(AssembliesToCmpress) will only contain assemblies that have changed since the last build.
+-->
+<Target Name="_CompressAssemblies"
+    DependsOnTargets="_CollectAssembliesToCompress"
+    Inputs="@(_AssembliesToCompress);@(_AndroidMSBuildAllProjects)"
```

You might also add:
```suggestion
    Inputs="@(_AssembliesToCompress);@(_AndroidMSBuildAllProjects);$(_AndroidBuildPropertiesCache)"
```
The `build.props` file has been the thing used to rerun all targets no matter what.